[week3] Spring Web ToDo REST API (2)

SUBIN606 opened the pull request:
안녕하세요. 이번 한 주도 잘 부탁 드립니다. 🙂

### 과제 목표
우리가 작성한 코드가 올바르게 동작하는지 자동으로 확인해 줄 수 있는 유닛 테스트를 작성해 주세요.
컨트롤러는 유닛 테스트뿐만 아니라 실제로 웹에서 동작하는 것처럼 작성하는 Spring Mock MVC 테스트를 작성해 주세요.

### ToDo
- [ ] HelloController 유닛 테스트
- [ ] HelloController MockMVC 테스트
- [x] Task 모델 테스트
- [x] TaskService 테스트
- [x] TaskController 유닛 테스트
- [x] TaskController MockMVC 테스트

### 요구사항 
- 모든 인수 테스트를 통과해야 합니다.
- 테스트 커버리지 100%를 달성해야 합니다.

SUBIN606 committed:
chore: update package-lock.json

SUBIN606 committed:
initial commit

2주차 까지 진행한 내용과 동일하게 맞췄습니다.

SUBIN606 commented:
질문이 있습니다.

1. 강의에서도 지난 번 까지 만든 ToDo API를 동일하게 사용하셔서 저도 2주차까지 만든 API를 그대로 이어서 진행하려 하는데요, 혹시 과제를 진행하는데 문제가 될까요? 
2. Task 모델 테스트라는게 `Task`클래스를 테스트 하는건가요?

SUBIN606 committed:
chore: 사용하지 않는 라이브러리 제거

SUBIN606 committed:
test: TaskCreateController mockMvc test

할 일 추가를 테스트 합니다.

hannut91 commented:
1. 지난 번에 사용한 ToDo API를 그래도 이어서 사용하셔도 됩니다.
2. 맞습니다.

SUBIN606 committed:
test: 테스트에 필요한 값 상수값으로 정의

SUBIN606 committed:
test: test TaskCreateService

SUBIN606 committed:
test: TaskReadService test

SUBIN606 committed:
test: TaskUpdateService test

SUBIN606 committed:
test: TaskDeleteService test

SUBIN606 committed:
test: TaskCreateServiceImplTest mocking으로 변경

SUBIN606 committed:
test: Task와 TaskDto test

SUBIN606 committed:
test: toString() 테스트 추가

SUBIN606 committed:
test: 예외처리 테스트 추가

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/tasks/TaskCreateControllerWebTest.java:
```diff
@@ -0,0 +1,55 @@
+    @Test
+    void validateTask() throws Exception {
+        mockMvc.perform(post("/tasks")
+                .content(objectMapper.writeValueAsString(INVALID_TASK_DTO))
```
할 일에 텍스트가 비어있는 경우를 테스트하고 있는 것 같아요그래서 `invalid`라는 단어를 사용하신 것 같아요. 그런데 유효하지 않다는 것은 더 많은 뜻을 담고 있습니다. 아예 null이 들어온다던가 혹은 할 일이 배열로 들어온다던가 하는 것들이 있죠.

너무 범용적인 의미를 가지고 있는 말은 사용하기가 편하지만, 그만큼 다르게 해석될 수 있습니다. 따라서 더 구체적인 말을 사용해 주세요.

hannut91 commented on app/src/test/java/com/codesoom/assignment/domains/TaskDtoTest.java:
```diff
@@ -0,0 +1,64 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+
+class TaskDtoTest {
```
`TaskDto` 객체에 대해서 굉장히 꼼꼼하게 작성해주신 것 같아요. 이 객체를 어떻게 사용하는지를 간단하게만 작성하는 것도 좋을 것 같아요

```java
TaskDto taskDto = new TaskDto();
taskDto.setTitle("책 읽기");

assertThat(taskDto.getTitle()).isEqualTo("책 읽기");

Task task = taskDto.toTask();

assertThat(task.getTitle()).isEqualTo("책 읽기");
```

hannut91 commented on app/src/test/java/com/codesoom/assignment/services/TaskCreateServiceImplTest.java:
```diff
@@ -0,0 +1,62 @@
+@ExtendWith(MockitoExtension.class)
+class TaskCreateServiceImplTest {
+
+    @InjectMocks
+    private TaskCreateServiceImpl service;
+
+    @Mock
+    private InMemoryTaskRepositoryImpl repository;
```
`@InjectMocks`와 `@Mock`은 어떤 차이가 있나요?

SUBIN606 replied:
- `@Mock`은 모의 클래스를 생성할 때 붙이는 애노테이션으로, `Mockito.mock()`으로 모의 클래스를 생성하는 방식을 더 간단하게 할 수 있도록 합니다.
- `@InjectMocks`는 의존성 주입을 필요로하는 필드에 붙이는 애노테이션입니다. `@Mock`이나 `@Spy` 애노테이션이 붙은 필드를 자동으로  의존성 주입 받을 수 있습니다. 생성자 주입, setter 주입, 필드 주입 방식을 지원합니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/services/TaskCreateServiceImplTest.java:
```diff
@@ -0,0 +1,62 @@
+        final TaskDto taskDto = new TaskDto(TASK_TITLE);
+
+        //when
+        given(repository.generateId()).willReturn(FIRST_ID);
```
테스트 준비를 위한 코드인 것 같아서, when에 해당하지는 않는 것 같아요.

hannut91 commented on app/src/test/java/com/codesoom/assignment/services/TaskUpdateServiceImplTest.java:
```diff
@@ -0,0 +1,69 @@
+
+    private static final Task BEFORE_TASK = new Task(1L, "old title");
+    private static final Task AFTER_TASK = BEFORE_TASK.updateTitle("new title");
+    private static final String INVALID_TASK_TITLE = "";
```
이 것도 `invalid`라기 보다는 `empty`라는 말이 더 어울릴 것 같아요

SUBIN606 committed:
test: 변수명 변경

INVALID를 EMPTY로 변경하여 더 구체적인 변수명을 사용합니다.

SUBIN606 committed:
test: TaskDtoTest 간단하게 수정

SUBIN606 committed:
test: response body가 없는 경우 테스트 추가

SUBIN606 committed:
test: AppTest

SUBIN606 committed:
test: TaskRepositoryTest

InMemoryTaskRepository를 테스트합니다.

SUBIN606 committed:
test: TaskCreateController unit test

SUBIN606 committed:
test: Task Controller tests

hannut91 commented on app/src/test/java/com/codesoom/assignment/repositories/InMemoryTaskRepositoryImplTest.java:
```diff
@@ -0,0 +1,87 @@
+    @AfterEach
+    void cleanup() {
+        repository.remove(TASK_ID);
+    }
```
테스트가 실행된 후, 메모리를 초기화하고 있는 것 같아요. 인메모리라서 상관이 없긴 하지만, 만약 중간에 테스트가 올바르게 동작하지 않아서 `cleanup`이 호출되지 않았을 때 다른 테스트에 영향이 있을 수 있어요. 따라서 테스트를 시작하기 전에 초기화를 하는게 더 좋을 것 같아요. 사실 둘 다 하는게 가장 안전하긴 합니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/repositories/InMemoryTaskRepositoryImplTest.java:
```diff
@@ -0,0 +1,87 @@
+            assertThat(id2 - id1).isEqualTo(1L);
+        });
+    }
+}
```
![image](https://user-images.githubusercontent.com/14071105/162027934-461fdcc4-37de-4cca-a095-af6a1622d3f7.png)

코드 마지막 줄에는 빈 줄이 하나가 추가되어야 합니다. 그렇지 않으면 위와 같이 GitHub에서 표시가 됩니다.

이 것이 문제가 되는 이유는 POSIX 표준에서 Line을 `개행 문자로 끝나는 0개 이상의 개행 문자가 아닌 문자들의 시퀀스`으로 정의하고 있기 때문입니다.

```jsx
// Hello
```

만약 위와 같이 빈 줄이 없는 코드에서 아래에 코드를 추가하게 되면 Git은 다음과 같이 표시합니다.

```jsx
- // Hello
+ // Hello
+ // World
```

저는 단순히 아래 한 줄만 추가했는데 왜 세줄이 변경되었다고 나오는 걸까요? 그 이유는 줄바꿈 문자 `\n`가 들어갔기 때문입니다.

```jsx
- // Hello
+ // Hello\n
+ // World
```

만약 미리 빈 줄을 하나 추가한 상태에서 아래에 새로운 코드를 작성할 경우에는 다음과 같이 보여집니다.

```jsx
+ // World
```

이미 `\n`이 추가되어 있기 때문이죠. 따라서 코드 마지막 줄에 빈 줄 하나를 항상 추가해주는게 좋습니다.

혹은 자동으로 추가할 수 있도록 설정하는 방법도 있습니다. 종립님님의 블로그 글을 참고해 주세요

See also
  - [파일-마지막에-자동으로-1개의-개행-문자-추가하기](https://johngrib.github.io/wiki/intellij/#파일-마지막에-자동으로-1개의-개행-문자-추가하기)

hannut91 commented on app/src/test/java/com/codesoom/assignment/services/TaskUpdateServiceImplTest.java:
```diff
@@ -0,0 +1,69 @@
+        given(repository.findById(BEFORE_TASK.getId())).willReturn(BEFORE_TASK);
+
+        //when
+        service.updateTaskById(BEFORE_TASK.getId(), new TaskDto("new title"));
```
`TaskDto`도 미리 선언해놓고 사용해도 괜찮을 것 같네요

hannut91 commented on app/src/test/java/com/codesoom/assignment/services/TaskUpdateServiceImplTest.java:
```diff
@@ -0,0 +1,69 @@
+    @Test
+    void thrownTaskInvalidException() {
+        //given
+        final TaskDto invalidTaskDto = new TaskDto(EMPTY_TASK_TITLE);
```
빈 문자열도 상수로 분리하신 것 같아요. 하지만 오히려 이해하는데 더 어려운 것 같아요. `new TaskDto("")` 이렇게 하면 그냥 빈 문자열이라는 것을 바로 알 수 있죠

SUBIN606 replied:
변경했습니다. 더 직관적인 것 같네요 😁

hannut91 commented on app/src/test/java/com/codesoom/assignment/services/TaskReadServiceImplTest.java:
```diff
@@ -0,0 +1,73 @@
+        //given
+        given(repository.findById(FIRST_TASK.getId())).willReturn(FIRST_TASK);
+
+        //when
+        final Task findTask = service.findTaskById(FIRST_TASK.getId());
+
+        //then
+        assertThat(findTask).isEqualTo(FIRST_TASK);
```
`findTask`라고 하면 task를 찾으라는 함수처럼 보일 것 같아요. 물론 앞에 타입이 있긴 하지만요. `task`는 어떨까요?

SUBIN606 replied:
넵 `find`로 시작하니까 함수처럼 보일 수 있겠네요. `task`로 변경했습니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/tasks/TaskUpdateControllerTest.java:
```diff
@@ -0,0 +1,67 @@
+                () -> controller.updateTask(NOT_EXIST_ID, new TaskDto("new title")));
+    }
+
+    @DisplayName("빈 제목으로 할 일을 수정하려고 하면 예외가 발생한다.")
+    @Test
+    void thrownInvalidFormatException() {
+        assertAll(() -> {
+            assertThrows(TaskInvalidFormatException.class,
+                    () -> controller.updateTask(NOT_EXIST_ID, new TaskDto(EMPTY_TASK_TITLE)));
+            assertThrows(TaskInvalidFormatException.class,
+                    () -> controller.updateTask(NOT_EXIST_ID, new TaskDto(BLANK_TASK_TITLE)));
+            assertThrows(TaskInvalidFormatException.class,
+                    () -> controller.updateTask(NOT_EXIST_ID, new TaskDto(null)));
+        });
+    }
```
할 일을 없데이트 할 수 없는 경우들을 테스트하고 싶으셨던 것 같아요. ParametrizedTest를 사용해보는 것은 어떨까요?

See also
  - https://junit.org/junit5/docs/5.3.0/api/org/junit/jupiter/params/ParameterizedTest.html

SUBIN606 replied:
`@ParameterizedTest`를 사용해봤습니다. 새로운 기능을 알게되어 좋네요!

그리고 궁금한 점이 하나 생겼는데요, `@ParameterizedTest`를 사용하면 보통 DisplayName은 어떻게 작성하나요?

<img src="https://user-images.githubusercontent.com/66243183/162149999-4a427a02-a700-4adb-80c8-ff46702adc95.png" width="400px"/>
현재는 이렇게 적어봤는데 더 좋은 방법이 떠오르지 않습니다.. 🙄

hannut91 replied:
`빈 제목`대신에 값을 보여주면 좋을 것 같네요. 만약 `@ParameterizedTest`가 의도가 잘 드러나지 않는 것 같아면 다음과 같이 해도 좋을 것 같아요. 안되는 경우가 `"", " ", null`이란 것을 한 번에 보여줄 수 있어서요.

```java
Arrays.asList("", " ", null).stream()
        .forEach(value ->
                assertThrows(TaskInvalidFormatException.class,
                        () -> controller.updateTask(NOT_EXIST_ID, new TaskDto(value)))
        );
```

SUBIN606 replied:
<img src="https://user-images.githubusercontent.com/66243183/162343279-d528ec3b-7b17-4396-bd44-84fa93d1be92.png" width="350px" />

`빈 제목` 대신 값을 보여주도록 수정하니 한결 나아졌네요. 👍

SUBIN606 committed:
test: 서비스 테스트 수정

SUBIN606 committed:
test: 컨트롤러 mockMvc테스트 아이디를 beforeEach로 받아오는 것으로 수정

SUBIN606 committed:
test: InMemoryRepository테스트 AfterEach 제거

SUBIN606 committed:
test: ParameterizedTest로 수정 할 수 없는 값 테스트

SUBIN606 commented on app/src/test/java/com/codesoom/assignment/controllers/tasks/TaskUpdateControllerWebTest.java:
```diff
@@ -0,0 +1,89 @@
+    private static final TaskDto EMPTY_TASK_DTO = new TaskDto("");
+    private static final TaskDto BLANK_TASK_DTO = new TaskDto("");
+
+    @BeforeEach
+    void setup() throws Exception {
+        final ResultActions actions = mockMvc.perform(post("/tasks")
+                .content(objectMapper.writeValueAsString(new TaskDto("title")))
+                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isCreated());
+
+        final Task task = objectMapper.readValue(actions.andReturn().getResponse().getContentAsString().getBytes(), Task.class);
+        this.EXIST_ID = task.getId();
+    }
```
테스트가 깨져서 `@BeforeEach`로 먼저 할 일을 하나 추가 하고, 추가된 ID를 사용하는 방식으로 변경해봤습니다. 
클래스 별로 테스트 코드를 돌리면 괜찮은데, 모든 테스트를 한 번에 돌리면 실패하더라구요.

제가 테스트 코드를 짜는게 처음이라 이렇게 해도 되는 건지 잘 모르겠네요..

hannut91 replied:
수정하는 테스트를 작성하기 위해서, 미리 데이터를 만들어 놓도록 해주셨군요. 수정하는 테스트가 할 일을 만드는 테스트에 의존하고 있는 부분이 조금 우려스럽지만, 이미 다른 테스트에서 할 일을 만드는 테스트를 충분히 작성하였고, 믿고 사용할 수 있다면 괜찮습니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/services/TaskUpdateServiceImplTest.java:
```diff
@@ -0,0 +1,69 @@
+    private TaskUpdateServiceImpl service;
+
+    @Mock
+    private InMemoryTaskRepositoryImpl repository;
```
인터페이스를 구현해서 `Impl`을 붙이신 것 같아요. 어떤 인터페이스를 구현한 객체가 이름을 붙이기가 애매해서 `Impl`을 붙인게 지금까지 계속 사용되어서 그런거에요. 즉 객체의 역할이 불분명한 코드인데, 객체가 하는 일에 대해서 고민해야 할 신호입니다. 이러한 이름은 절대 권장하지 않습니다.

범용적인 이름일수록 사용할 땐 편하지만, 읽을 때는 고통스럽습니다. 그냥 `InMemoryTaskRepository`가 맞습니다.

SUBIN606 committed:
refactor: 할 일 추가, 수정 메서드 변경

- InMemoryTaskRepository의 generateId()를 public에서 private으로 변경해
저장될 때 알아서 id가 증가되도록 합니다.
- 할 일을 추가, 수정 할 때 리포지토리에서 결과를 반환하도록 변경합니다.

SUBIN606 committed:
test: 유효성 검증 테스트를 반복문으로 변경

SUBIN606 committed:
test: getTasks 테스트 추가

SUBIN606 committed:
refactor: 테스트 코드 정리

SUBIN606 committed:
refactor: 사용하지 않는 변수 정리

SUBIN606 committed:
test: TaskDto변경에 따른 수정

SUBIN606 committed:
fix: 할 일을 수정하지 못하면 Null을 반환하도록 함

InMemoryRepository에서 할 일을 찾지 못해 수정하지 못하면
null을 반환하게 변경합니다.

SUBIN606 committed:
test: InMemoryRepositoryTest를 계층형 구조로 작성

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/ControllerAdvice.java:
```diff
@@ -0,0 +1,38 @@
+    protected ResponseEntity<ErrorResponse> handleNoSuchElementException(TaskNotFoundException e) {
+        final ErrorResponse errorResponse = new ErrorResponse(e.getCode(), e.getMessage());
+        return new ResponseEntity<>(errorResponse, HttpStatus.NOT_FOUND);
+    }
```
여기에도 `ResponseEntity`대신에 `@ResponseStatus` 를 쓰면 좋을 것 같아요

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/tasks/TaskDeleteController.java:
```diff
@@ -0,0 +1,28 @@
+
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @DeleteMapping("/{id}")
+    public void deleteTask(@PathVariable(name = "id") Long id) {
```
path variable과 변수명이 같을 경우 `(name = "id")`를 생략할 수 있는데, 명시적으로 작성하는게 더 좋을까요?

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/tasks/TaskReadController.java:
```diff
@@ -0,0 +1,33 @@
+
+    private final TaskReadService service;
+
+
```
여기 코드가 두 줄 띄어져 있네요

hannut91 commented on app/src/main/java/com/codesoom/assignment/domains/Task.java:
```diff
@@ -0,0 +1,41 @@
+        return this;
+    }
+
+    @Override
+    public String toString() {
+        return "Task{" +
+                "id=" + id +
+                ", title='" + title + '\'' +
+                '}';
+    }
```
`String.format`을 사용해 보는 것은 어떨까요?

hannut91 commented on app/src/main/java/com/codesoom/assignment/exceptions/TaskInvalidFormatException.java:
```diff
@@ -0,0 +1,24 @@
+import org.springframework.http.HttpStatus;
+
+/** 유효하지 않은 형식의 할 일 추가 요청 시 발생하는 exception */
+public class TaskInvalidFormatException extends RuntimeException{
```
유효하지 않다는 말도 너무 범용적이라서 해석될 수 있는 여지가 너무 많을 것 같아요.

hannut91 commented on app/src/main/java/com/codesoom/assignment/exceptions/TaskInvalidFormatException.java:
```diff
@@ -0,0 +1,24 @@
+
+    private String code;
+
+    public TaskInvalidFormatException(String message) {
+        super(message);
+        this.code = String.valueOf(HttpStatus.BAD_REQUEST.value());
+    }
+
+    public TaskInvalidFormatException(String code, String message) {
+        super(message);
+        this.code = code;
+    }
```
```java
private String code;

public TaskInvalidFormatException(String code, String message) {
    super(message);
    this.code = code;
}

public TaskInvalidFormatException(String message) {
    this(String.valueOf(HttpStatus.BAD_REQUEST.value()), message);
}
```

주 생성자와 부 생성자를 나누어서, 객체를 초기화 하는 일은 하나의 생성자가 하게 하고, 다른 생성자는 객체를 만들 수 있는 다양한 방법을 지원하도록 하여, 주 생성자가 해야할 일을 하게 하고, 부 생성자는 주 생성자로 전달할 일만 하게하여 중복을 제거할 수 있습니다.

> '하나의 주 생성자와 다수의 부 생성자 원칙(one primay, many secondary)의 핵심은 무엇일까요? 이 원칙이 중복 코드를 방지하고 설계를 더 간결하게 만들기 때문에 유지보수성이 향상된다는 점입니다. - 엘레강트 오브젝트 27p

흠 생각해보니 코틀린은 기본 값을 할당할 수 있어서 더 간단해지겠네요.

```kotlin
class TaskInvalidFormatException(
    val message,
    val code = String.valueOf(HttpStatus.BAD_REQUEST.value()),
): RuntimeException(message) {
    // ...
}
```

SUBIN606 replied:
한 번도 생각해 보지 못한 부분이었네요.. 해당 개념에 대해서는 더 공부해보도록 하겠습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/repositories/InMemoryTaskRepository.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    /** id와 매핑되는 할 일을 수정 후 변경 결과를 반환합니다. */
+    @Override
+    public Task update(Long id, Task updatedTask) {
+        tasks.replace(id, updatedTask);
+        return tasks.replace(id, updatedTask) == null ? null : updatedTask;
+    }
```
함수가 null을 반환할 수 있군요. 이 함수를 사용하는 입장에서는 null을 처리하지 않을경우 NPE가 발생할 수 있고, 이를 강제할 수 있는 방법이 없네요. 코틀린은 null을 허용하는 타입을 사용해서 컴파일 타임에 이러한 부분을 처리하지 않았다고 에러를 알려주지만, 자바는 실행하기 전까지는 파악할 수 없기 때문에 자칫 잘못하면 큰 일이 일어날 수 있어요. 따라서 null을 반환하는 것은 좋지 않습니다.

SUBIN606 replied:
`TaskUpdateServiceImpl`에서 막아주고 있어서 null을 반환해도 된다고 생각했네요. 수정해보겠습니다!

hannut91 commented on app/src/main/java/com/codesoom/assignment/services/TaskCreateServiceImpl.java:
```diff
@@ -0,0 +1,40 @@
+        return repository.save(taskDto.toTask());
+    }
+
+    /**
+     * 사용자가 입력한 데이터의 유효성을 검증합니다.
+     *
+     * @param taskDto 사용자가 입력한 데이터
+     * @throws TaskInvalidFormatException
+     *         데이터의 유효성 검증을 통과하지 못한 경우
+     */
```
이 함수에서 하는 일을 잘 작성해주신 것 같아요. 이 함수가 그 이후에 어떤 일을 하는지 더 명시해주면 어떨까요? 예를들어서 `사용자가 입력한 데이터의 유효성을 검증합니다. 만약 검증에 실패하면 예외를 던집니다.`는 어떨까요? 반환타입이 void라서 예상하기가 어려워 추가적으로 설명해주면 어떨까 생각해 보았어요.

SUBIN606 replied:
넵 javaDoc에 작성해 두니까 함수를 바로 이해할 수 있어 좋네요!

SUBIN606 committed:
refactor: ResponseStatus애노테이션으로 응답 상태값 설정

SUBIN606 committed:
refactor: PathVariable name생략

SUBIN606 committed:
fix: 할 일 수정 시 null 반환 제거

SUBIN606 committed:
refactor: 주 생성자와 부 생성자로 분리

SUBIN606 committed:
test: InMemoryRepository 테스트 수정

hannut91 commented:
테스트 코드를 처음 짜보셨군요! 굉장히 잘해주셔서 몰랐어요. 앞으로도 남은기간 더 많은 연습을 통해 자신의 것으로 만드셨으면 좋겠습니다. 이번 한 주도 고생 많으셨습니다!
